wrapper.ts

```diff
@@ -1,360 +1,159 @@
-          };
-        };
-      }
-
-      singleCallback = function (kind, data) {
-        if (kind === 'source') {
-          return readCallback(data);
-        } else if (kind === 'smt-query') {
-          return smtSolverCallback(data);
-        } else {
-          assert(false, 'Invalid callback kind specified.');
+      response.pipe(memoryStream);
+      response.on('end', () => {
+        // Based on the require-from-string package.
+        const soljson = new Module();
+        soljson._compile(memoryStream.toString(), `soljson-${versionString}.js`);
+
+        if (module.parent && module.parent.children) {
+          // Make sure the module is plugged into the hierarchy correctly to have parent
+          // properly garbage collected.
+          module.parent.children.splice(module.parent.children.indexOf(soljson), 1);
         }
-      };
 
-      singleCallback = wrapCallbackWithKind(singleCallback);
-    } else {
-      // Old Solidity version only supported imports.
-      singleCallback = wrapCallback(readCallback);
-    }
-
-    // This is to support multiple versions of Emscripten.
-    const addFunction = soljson.addFunction || soljson.Runtime.addFunction;
-    const removeFunction = soljson.removeFunction || soljson.Runtime.removeFunction;
-
-    const cb = addFunction(singleCallback, 'viiiii');
-    let output;
-    try {
-      args.push(cb);
-      if (isVersion6) {
-        // Callback context.
-        args.push(null);
-      }
-      output = compile.apply(undefined, args);
-    } catch (e) {
-      removeFunction(cb);
-      throw e;
-    }
-    removeFunction(cb);
-    if (reset) {
-      // Explicitly free memory.
-      //
-      // NOTE: cwrap() of "compile" will copy the returned pointer into a
-      //       Javascript string and it is not possible to call free() on it.
-      //       reset() however will clear up all allocations.
-      reset();
+        callback(null, wrapper(soljson.exports));
+      });
     }
-    return output;
-  };
+  }).on('error', function (error) {
+    callback(error);
+  });
+}
 
-  let compileJSON = null;
-  if ('_compileJSON' in soljson) {
-    // input (text), optimize (bool) -> output (jsontext)
-    compileJSON = soljson.cwrap('compileJSON', 'string', ['string', 'number']);
+// Expects a Standard JSON I/O but supports old compilers
+function compileStandardWrapper (compile, inputRaw, readCallback) {
+  if (!isNil(compile.compileStandard)) {
+    return compile.compileStandard(inputRaw, readCallback);
   }
 
-  let compileJSONMulti = null;
-  if ('_compileJSONMulti' in soljson) {
-    // input (jsontext), optimize (bool) -> output (jsontext)
-    compileJSONMulti = soljson.cwrap('compileJSONMulti', 'string', ['string', 'number']);
-  }
+  let input: { language: string, sources: any[], settings: any };
 
-  let compileJSONCallback = null;
-  if ('_compileJSONCallback' in soljson) {
-    // input (jsontext), optimize (bool), callback (ptr) -> output (jsontext)
```

Comment:
The extra info from these comments (i.e. that the input and output is meant to be JSON) seems lost, even in the typed version.

Reply:
I did not feel them a need since they would be explicitly typed in the following PR but I will add them back anyway.

Reply:
done

Reply:
I mean, there isn't a type for JSON is there? It's just a string. The rest is there but the fact that input and output JSON should be kept somewhere in the docstring.

Reply:
yeah I have added them